### PR TITLE
changes for python 3.9 and python 3.10

### PR DESCRIPTION
**Pull request by tzukpolinsky**
Hi
I started working with your frame work as part of my role as data scientist at Klavir lab in Haifa university, i have to say, you frame work is really amazing for the amount of analysis and functionality it enables :)

So some package version needed to be upgraded, and in therefor some code as well.

I created ROI's and model on my data, and it seems all the outpout is as it should be and no unnecessary  exception happends.

Hope this contribute to your project:)

Cheers

### CODE REVIEW

**Comment by tzukpolinsky** [comment]
Hi
I will check this out and let you know:)

With regards
Tzuk Polinsky

**Comment by tzukpolinsky** [comment]
Hi
tested for creating project and running some functionality, didnt trained a full model just some simple analytical from the vast variety in your app.

Note:
project that is originally create with python 3.8+ are not compatible for python 3.7 and 3.6 because of the different pickle protocol in the advanced versions (protocol 5 vs 4).

fixed some crashes from tinker,  just some annoying stuff with the utf it likes and dont likes

Hoped i helped in any way:)

**Comment by sronilsson** [comment]
Thanks @tzukpolinsky! It does, give me a little time and I will merge this

Simon

File: simba/mixins/train_model_mixin.py
Comment (sronilsson) [inline simba/mixins/train_model_mixin.py:1]: Cool! To play it safe, maybe: 

        if hasattr(clf, 'n_features_'):
            clf_n_features = clf.n_features_
        elif hasattr(clf, 'n_features_in_'):
            clf_n_features = clf.n_features_in_
        else:
            raise InvalidInputError(msg=f'Could not determine the number of features features in the classifier {model_name}', source=self.__class__.__name__) 
.....

File: simba/utils/warnings.py
Comment (sronilsson) [inline simba/utils/warnings.py:1]: i like this one!

File: simba/mixins/train_model_mixin.py
Comment (tzukpolinsky) [reply simba/mixins/train_model_mixin.py:1]: My opinion that playing safe in such a large project just makes it difficult and misleading in later development:)

But obviously you are the boss so what ever you say and I would change it:)

File: simba/utils/warnings.py
Comment (tzukpolinsky) [reply simba/utils/warnings.py:1]: thank you:)

File: simba/SimBA.py
Comment (sronilsson) [inline simba/SimBA.py:1]: 👍🏻 👍🏻

File: simba/roi_tools/ROI_define.py
Comment (sronilsson) [inline simba/roi_tools/ROI_define.py:1]: I can see how this is more reliable but I am hitting an error when I try to launch on MacOS, have you seen this before? 

QObject::setParent: Cannot set parent, new parent is in a different thread
QPixmap: It is not safe to use pixmaps outside the GUI thread
zsh: segmentation fault  python App.py

**Comment by tzukpolinsky** [comment]
Its around the qt version
Maybe updating the qt5..?
Sudo apt-get upgrade qt5-dev

**Comment by tzukpolinsky** [comment]
Hi
can we resolve the pull request?:)

i would like to merge some change to my main and I dont want to burden this one with it

**Comment by sronilsson** [comment]
Yes sorry for delay, but meaning to each day and slipped. I'm back at the computer tonight and will fix that single conflict and merge.

**Comment by sronilsson** [comment]
@tzukpolinsky - I don't have many tests.. would love more. But I realized two tests are failing:

Both appear pandas related:

``
AttributeError: 'DataFrame' object has no attribute '_is_numeric_mixed_type’. 
``

For that one I could just commend out the assertion if `_is_numeric_mixed_type` isn’t a thing in newer pandas, or maybe the test returns a df and it was previously a series in latest pandas?

And:

`df = pd.read_excel(file_path, sheet_name=None,usecols=EXPECTED_FIELDS).popitem(last=False)[1]
TypeError: popitem() takes no keyword arguments` Which I’ve never seen before.. seems latest pandas dropped the `last` argument?


If I pin pandas==0.25.3 in requirements.txt it worked though. Is it essential that pandas is the latest version?

**Comment by tzukpolinsky** [comment]
Hi

I can try and solve this tomorrow evening

The version becomes deprecated for higher version of python, so i would
recommend to try and work with the newest

Tomorrow i would give it a go

**Comment by sronilsson** [comment]
Thank you, the first assertion is not important we can live without that. 

The second error though, on `df = pd.read_excel(file_path, sheet_name=None,usecols=EXPECTED_FIELDS).popitem(last=False)[1]` needs to be replaced to work in python3.9. 

It is used when people have Noldus Observer hand-annotations in excel workbooks. An Excel workbook from Noldus Observer contains many spreadsheets for a single video, and most spreadsheets can be discarded, and we only want one of them [HERE](https://github.com/sgoldenlab/simba/blob/d45e63c94b2b5ef95cfdd61ee7a1ac15ae35293c/simba/third_party_label_appenders/observer_importer.py#L92C21-L92C21).

**Comment by sronilsson** [comment]
Ah it should be an easy fix, I think this has something to do with it from chatgpt:

"""
**key-value pairs in dictionaries was not guaranteed prior to Python 3.7. Starting from Python 3.7, dictionaries maintain the insertion order, and this order is preserved when using popitem(). However, dictionaries in earlier Python versions do not maintain order, and the behavior of popitem() may be less predictable in those versions.**
"""

**Comment by tzukpolinsky** [comment]
will the following replacment works?
```df = next(iter(pd.read_excel(file_path, sheet_name=None, usecols=EXPECTED_FIELDS).items()))[1]
```

sorry i dont have Noldus data to check it

the idea is to convert the pandas dic to an iterable and than trying and get the first value.

**Comment by sronilsson** [comment]
Yes I will figure it out! Thanks again @tzukpolinsky this is very helpful and useful. 

FYI: I have two installation tests that runs every time the main branch updates. [One](https://github.com/sgoldenlab/simba/blob/master/.github/workflows/tests_py36.yml) installs SimBA on python3.6, and runs some tests and blacks the code. I just inserted a [second](https://github.com/sgoldenlab/simba/blob/master/.github/workflows/tests_py310.yml) that installs SimBA on python3.10 (but doesn't perform any tests as yet). 

The trouble I had is that those two install tests use the same requirement.txt, but pandas==0.25.3 will not install on python3.10, and latest pandas version will not install on python3.6 (you need python >= 3.9). 

For now, I inserted [markers](https://peps.python.org/pep-0508/#environment-markers) in the requirement.txt, so pandas dependency is different depending on python version: 

image

A bit hacky but at least the code will install in both environment..














[THIS](https://peps.python.org/pep-0508/#environment-markers) is a possible solution for the requirements. I

**Comment by tzukpolinsky** [comment]
looks good:)

I have few more changes and functionality I did at my fork (haven't push them yet) we could discussed them over emails, if you would like:)

glad I could help

**Pull request merged**